jwtmiddleware.go

```diff
@@ -84,38 +103,116 @@ func (m *JWTMiddleware) Handler(h http.Handler) http.Handler {
 	})
 }
 
-func (m *JWTMiddleware) CheckJWT(w http.ResponseWriter, r *http.Request) error {
-
+// FromAuthHeader is a "TokenExtractor" that takes a give request and extracts
+// the JWT token from the Authorization header.
+func FromAuthHeader(r *http.Request) (string, error) {
 	authHeader := r.Header.Get("Authorization")
 	if authHeader == "" {
-		if m.Options.CredentialsOptional {
-			return nil
-		}
-		errorMsg := "Authorization header isn't sent"
-		m.Options.ErrorHandler(w, r, errorMsg)
-		return errors.New(errorMsg)
+		return "", nil // No error, just no token
 	}
 
+	// TODO: Make this a bit more robust, parsing-wise
 	authHeaderParts := strings.Split(authHeader, " ")
 	if len(authHeaderParts) != 2 || strings.ToLower(authHeaderParts[0]) != "bearer" {
-		errorMsg := "Authorization header format must be Bearer {token}"
-		m.Options.ErrorHandler(w, r, errorMsg)
-		return errors.New(errorMsg)
+		return "", fmt.Errorf("Authorization header format must be Bearer {token}")
+	}
+
+	return authHeaderParts[1], nil
+}
+
+// FromParameter returns a function that extracts the token from the specified
+// query string parameter
+func FromParameter(param string) TokenExtractor {
+	return func(r *http.Request) (string, error) {
+		return r.URL.Query().Get(param), nil
+	}
+}
+
+// FromFirst returns a function that runs multiple token extractors and takes the
+// first token it finds
+func FromFirst(extractors ...TokenExtractor) TokenExtractor {
+	return func(r *http.Request) (string, error) {
+		for _, ex := range extractors {
+			token, err := ex(r)
+			if err != nil {
+				return "", err
+			}
+			if token != "" {
+				return token, nil
+			}
+		}
+		return "", nil
+	}
+}
+
+func (m *JWTMiddleware) CheckJWT(w http.ResponseWriter, r *http.Request) error {
+	// Use the specified token extractor to extract a token from the request
+	token, err := m.Options.Extractor(r)
+
+	// If debugging is turned on, log the outcome
+	if m.Options.Debug {
+		if err != nil {
+			log.Printf("Error extracting JWT: %v", err)
+		} else {
+			log.Printf("Token extracted: %s", token)
+		}
 	}
 
-	token := authHeaderParts[1]
+	// If an error occurs, call the error handler and return an error
+	if err != nil {
+		m.Options.ErrorHandler(w, r, err.Error())
+		return fmt.Errorf("Error extracting token: %v", err)
+	}
+
+	// If the token is empty...
+	if token == "" {
+		// Check if it was required
+		if m.Options.CredentialsOptional {
+			if m.Options.Debug {
+				log.Printf("  No credentials found (CredentialsOptional=true)")
+			}
+			// No error, just no token (and that is ok given that CredentialsOptional is true)
+			return nil
+		}
 
+		// If we get here, the required token is missing
+		errorMsg := "Required authorization token not found"
+		m.Options.ErrorHandler(w, r, errorMsg)
+		if m.Options.Debug {
```

Comment:
We're doing this m.Options.Debug everywhere.

I think we should extract it to a function like:

```go
LogIfDebug(m.Options, "Error parsing token: %v", err);
```

This way code will look more compact and better and not have the logging flag everywhere.

What do you think?